controllers/oauth.controller.js

```diff
@@ -0,0 +1,171 @@
+      const oauth = new Oauth({ token: userToken });
+      await oauth.save();
+      res.status(200).send({ message: userToken });
+    } else {
+      res.status(401).send("User unauthorized!!!");
+    }
+  } catch (error) {
+    if (error) {
+      res.status(401).send({ message: "User unauthorized" });
+    } else {
+      logger.error("Could not login: ", error);
+      res.status(500).send({ message: "Internal Server Error" });
+    }
+  }
+  return null;
+};
+
+const restPassword = async (req, res) => {
+  try {
+    const { error, value } = patchUserPasswordSchema.validate(req.body);
+    if (error) {
+      return res.status(400).send({ message: error.message });
+    }
+    const authHeader = req.headers.authorization;
+    const verifyToken = jwt.verify(authHeader, process.env.TOKEN_KEY);
+    // eslint-disable-next-line no-underscore-dangle
+    const userPassword = await User.findOne({ _id: verifyToken._id });
+    value.newPassword = await bcrypt.hash(value.newPassword, 10);
+    value.currentPassword = await bcrypt.compare(
+      value.currentPassword,
+      userPassword.password,
+    );
+    if (value.currentPassword === true) {
+      userPassword.password = value.newPassword;
+      await userPassword.save();
+      res.status(200).send();
+    } else {
+      res.status(400).send({ message: "Invalid Password details" });
+    }
+  } catch (error) {
+    if (error) {
+      res.status(400).send({ message: "Invalid token" });
+    } else {
+      logger.error("Could not update password: ", error);
+      res.status(500).send({ message: "Internal Server Error" });
+    }
+  }
+  return null;
+};
+
+const sendEmail = async (req, res) => {
+  try {
+    const { error, value } = postOtpSchema.validate(req.body);
+    if (error) {
+      return res.status(400).send({ message: error.message });
+    }
+    const userOtp = parseInt(Math.random() * 1000000, 10);
+    const expireOtp = new Date().getTime() + 600 * 1000;
+    const user = await User.findOne(req.body);
+
+    if (user.email === value.email) {
+      const otp = new Otp({
+        email: value.email,
+        otp: userOtp,
+        expiresIn: expireOtp,
+      });
+
+      const transporter = nodemailer.createTransport(
+        {
+          host: process.env.HOST,
+          service: process.env.SERVICE,
+          secure: true,
+
+          auth: {
+            user: process.env.USERMAIL,
+            pass: process.env.PASSWORD,
+          },
+        },
+      );
+      const mailOptions = {
+        from: process.env.USERMAIL,
+        to: value.email,
+        subject: "forget password",
+        text: `Dear user your one time security code is:${userOtp}`,
+      };
+      await transporter.sendMail(mailOptions, (err) => {
+        if (err) {
+          res.status(400).send({ message: "Otp coulden't send" });
+        } else {
+          res.status(201).send({ message: "Otp send successfully" });
+        }
+      });
+      await otp.save();
```

Comment:
We should first save otp and then send it. It doesn't make sense if we couldn't save it after sending it to email, It won't work.